### ISSUE TITLE
Feature request: Document guidance for installing packages not available in selected channels

### ISSUE THREAD

**Issue by joooeey** [mamba-org/mamba#2479]
### Troubleshooting docs

- [X] My problem is not solved in the Troubleshooting docs

### How did you install Mamba?

Mambaforge

### Search tried in issue tracker and docs

Search terms: "pypi" and "pip". I only found #342 which is related but not about docs, and #1899 which doesn't give guidance how to do this in the console.

### Latest version of Mamba

- [X] My problem is not solved with the latest version

### Tried in Conda?

I have this problem with Conda as well, without using Mamba

### Describe your issue

After trying to install a package that is not available from `conda-forge` but from `pip` I have no idea what the next best course of action is. There are lots of opinions about this on the conda issue tracker and stackoverflow but no summary of drawbacks and benefits.

Since PyPI is the canonical package repository of the Python Software Foundation, such guidance should be provided, at least here, or preferably in the docs. This should be faster to accomplish than solving #342 and #1339.

### mamba info / micromamba info

```shell
$ mamba info

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.1) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


     active environment : del
    active env location : /home/lukas/mambaforge/envs/del
            shell level : 2
       user config file : /home/lukas/.condarc
 populated config files : /home/lukas/mambaforge/.condarc
          conda version : 23.1.0
    conda-build version : not installed
         python version : 3.10.10.final.0
       virtual packages : __archspec=1=x86_64
                          __cuda=11.7=0
                          __glibc=2.31=0
                          __linux=5.4.0=0
                          __unix=0=0
       base environment : /home/lukas/mambaforge  (writable)
      conda av data dir : /home/lukas/mambaforge/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /home/lukas/mambaforge/pkgs
                          /home/lukas/.conda/pkgs
       envs directories : /home/lukas/mambaforge/envs
                          /home/lukas/.conda/envs
               platform : linux-64
             user-agent : conda/23.1.0 requests/2.28.2 CPython/3.10.10 Linux/5.4.0-135-generic linuxmint/20.1 glibc/2.31
                UID:GID : 1000:1003
             netrc file : /home/lukas/.netrc
           offline mode : False
```

### Logs

```shell
$ mamba install "bleak>=0.20.2" -v

                  __    __    __    __
                 /  \  /  \  /  \  /  \
                /    \/    \/    \/    \
███████████████/  /██/  /██/  /██/  /████████████████████████
              /  / \   / \   / \   / \  \____
             /  /   \_/   \_/   \_/   \    o \__,
            / _/                       \_____/  `
            |/
        ███╗   ███╗ █████╗ ███╗   ███╗██████╗  █████╗
        ████╗ ████║██╔══██╗████╗ ████║██╔══██╗██╔══██╗
        ██╔████╔██║███████║██╔████╔██║██████╔╝███████║
        ██║╚██╔╝██║██╔══██║██║╚██╔╝██║██╔══██╗██╔══██║
        ██║ ╚═╝ ██║██║  ██║██║ ╚═╝ ██║██████╔╝██║  ██║
        ╚═╝     ╚═╝╚═╝  ╚═╝╚═╝     ╚═╝╚═════╝ ╚═╝  ╚═╝

        mamba (1.4.1) supported by @QuantStack

        GitHub:  https://github.com/mamba-org/mamba
        Twitter: https://twitter.com/QuantStack

█████████████████████████████████████████████████████████████


Looking for: ["bleak[version='>=0.20.2']"]

info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/ld_impl_linux-64-2.40-h41732ed_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/openssl-3.1.0-h0b41bf4_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/tk-8.6.12-h27826a3_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/wheel-0.40.0-pyhd8ed1ab_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/ca-certificates-2022.12.7-ha878542_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libsqlite-3.40.0-h753d276_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libgomp-12.2.0-h65d4601_19.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/tzdata-2023c-h71feb2d_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/bzip2-1.0.8-h7f98852_4.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/python-3.11.3-h2755cc3_0_cpython.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/_libgcc_mutex-0.1-conda_forge.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/readline-8.2-h8228510_1.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libgcc-ng-12.2.0-h65d4601_19.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libnsl-2.0.0-h7f98852_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libffi-3.4.2-h7f98852_5.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libexpat-2.5.0-hcb278e6_1.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/pip-23.1-pyhd8ed1ab_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/ncurses-6.3-h27087fc_1.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libuuid-2.38.1-h0b41bf4_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/libzlib-1.2.13-h166bdaf_4.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/setuptools-67.6.1-pyhd8ed1ab_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/xz-5.2.6-h166bdaf_0.json"
info     libmamba Loading single package record: "/home/lukas/mambaforge/envs/del/conda-meta/_openmp_mutex-4.5-2_gnu.json"
info     libmamba Reading cache files '/tmp/tmpbn7zkf0u.*' for repo index 'installed'
info     libmamba Searching index cache file for repo 'https://conda.anaconda.org/conda-forge/linux-64/repodata.json'
conda-forge/linux-64                                        Using cache
info     libmamba Searching index cache file for repo 'https://conda.anaconda.org/conda-forge/noarch/repodata.json'
conda-forge/noarch                                          Using cache
info     libmamba All targets to download are cached
Channel: conda-forge[linux-64,noarch], platform: linux-64, prio: 1 : 0
Cache path:  /home/lukas/mambaforge/pkgs/cache/497deca9.solv
info     libmamba Reading cache files '/home/lukas/mambaforge/pkgs/cache/497deca9.solv.*' for repo index 'https://conda.anaconda.org/conda-forge/linux-64'
info     libmamba Attempt load from solv "/home/lukas/mambaforge/pkgs/cache/497deca9.solv"
info     libmamba Metadata solv file: https://conda.anaconda.org/conda-forge/linux-64/repodata.json 1 W/"baa9173c33eb230888e658d51d654f4e" Fri, 21 Apr 2023 10:38:08 GMT 1.1_0.7.23
info     libmamba Metadata from SOLV are valid
Channel: conda-forge[linux-64,noarch], platform: noarch, prio: 1 : 0
Cache path:  /home/lukas/mambaforge/pkgs/cache/09cdf8bf.solv
info     libmamba Reading cache files '/home/lukas/mambaforge/pkgs/cache/09cdf8bf.solv.*' for repo index 'https://conda.anaconda.org/conda-forge/noarch'
info     libmamba Attempt load from solv "/home/lukas/mambaforge/pkgs/cache/09cdf8bf.solv"
info     libmamba Metadata solv file: https://conda.anaconda.org/conda-forge/noarch/repodata.json 1 W/"4ec9596b707721cc56f211ba6c41b70d" Fri, 21 Apr 2023 10:08:16 GMT 1.1_0.7.23
info     libmamba Metadata from SOLV are valid
info     libmamba Parsing MatchSpec bleak[version='>=0.20.2']
info     libmamba Parsing MatchSpec bleak[version='>=0.20.2']
info     libmamba Adding job: bleak >=0.20.2
info     libmamba Parsing MatchSpec ca-certificates
info     libmamba Parsing MatchSpec ca-certificates
info     libmamba Parsing MatchSpec openssl
info     libmamba Parsing MatchSpec openssl
info     libmamba Parsing MatchSpec python 3.11.*

Pinned packages:
  - python 3.11.*


info     libmamba Problem count: 1
info     libmamba Parsing MatchSpec bleak >=0.20.2
info     libmamba Parsing MatchSpec bleak >=0.20.2
Could not solve for environment specs
The following package could not be installed
└─ bleak >=0.20.2  does not exist (perhaps a typo or a missing channel).
info     libmamba Freeing solver.
info     libmamba Freeing pool.
```

### ~/.condarc

```yaml
channels:
- conda-forge
```

**Comment by joooeey** [comment]
My current soltution is the following:
```
mamba activate del
pip install "bleak>=0.20.2"
```

It works so far.

**Comment by jonashaag** [comment]
You can add it to conda-forge: https://conda-forge.org/docs/maintainer/adding_pkgs.html

**Comment by joooeey** [comment]
What benefits do I (and others) get from adding a package to conda-forge instead of just installing into the environment with `pip`? Given this is many of orders of magnitude more work. I've heard there are risks with using `pip` together with `conda` but I've never seen a specific explanation what these risks are and have no idea to what extent the risks are the same with `mamba`.

**Comment by JohanMabille** [comment]
Pip packages vendor their native dependencies (when they have ones). So you may end with different versions of a same library in your environment. This can lead to weird behavior in some cases (openssl for instance). Therefore it is usually recommended ot not mix pip packages with conda/mamba packages.

**Comment by mfhepp** [comment]
In a world where more and more of our code is running inside Docker containers, I think the strict recommendation to try to get all dependencies from conda-forge is no longer valid. The key reason for avoiding the mix of `pip` and `conda` components was that you might mess up your environment due to version conflicts not visible to `conda`. 

If everything is in a Docker container, well, in the worst of cases I will re-build that container, so I can take a more relaxed approach to mixing `pip` and `conda-forge` components or even install everything from `pip` inside a `conda` environment.

The advantage is that I save dozens of hours finding out which dependency is not yet updated on `conda-forge` and stops me from using a newer version of a core component.

Also, in the era of massive [Supply Chain Attacks](https://www.microsoft.com/en-us/securityengineering/opensource/ossthreats), the complex and, to the casual user, intransparent  ecosystem for `conda-forge` looks like an invitation for trouble to me.

**Comment by jonashaag** [comment]
conda-forge is actually much safer because each new package and each new version of a package goes through a public pull request whereas on PyPI there is no such audit trail that everyone can see. Things like typosquatting are much less probable on conda-forge than on PyPI.

**Comment by mfhepp** [comment]
> conda-forge is actually much safer because each new package and each new version of a package goes through a public pull request whereas on PyPI there is no such audit trail that everyone can see. Things like typosquatting are much less probable on conda-forge than on PyPI.

This may hold for pure conda-forge packages. But for many PyPi packages, you have basically two sources of malicious injections: From the (admittedly: flawed) PyPi ecosystem **and** from the conda-forge workflows and ecosystem, in particular, if independent developers create or maintain the conda feedstock or other components.

It is really easy to introduce really bad stuff into e.g. setup.py scripts and the large amounts of boilerplate code in scripts and "best practices" recipes with complex CI workflows are an invitation for trouble.

Basically I think that as of 2023, you cannot responsively execute Python code with external dependencies on a local machine without some substantial sandboxing (e.g. at least in a container with limited access rights). That holds for both conda-forge an PyPi (and other programming language ecosystems as well; Node.js is equally bad if not worse).

**Comment by jonashaag** [comment]
I think what you say is true for any package manager that doesn’t do sandboxing.

**Comment by mfhepp** [comment]
> I think what you say is true for any package manager that doesn’t do sandboxing. If you don’t trust other people, you have to write all the code yourself, including the compiler etc

I think that is too bold and too simplistic a statement. Of course, any division of labor and relying on others' code creates some risks, and none of us wants to write everything from scratch. **But Supply Chain Attacks are an issue**, in particular in the Python ecosystem, where

1. a complete programming language **with typically full access to a user's machine,**
2. a community with many people developing code who are no formally trained software engineers (admittedly including myself, but think of the many Data Science practitioners), and
3. **complexity and limited transparency**, as e.g. in the form of sophisticated CI workflows for publication of Python packages,

come together in one place.

And yes, PyPi/`pip` is no better than `conda-forge`/`conda`/`mamba`, just different. But the non-trivial packaging process and the frequent divergence between package developer and maintainer on `conda-forge` are disasters waiting to happen.

**Comment by jonashaag** [comment]
I think this discussion is important to have, but this bug tracker isn't the right place.
